src/libcore/iter.rs

```diff
@@ -3270,6 +3270,50 @@ impl<A, B> DoubleEndedIterator for Zip<A, B> where
 ///
 /// [`map()`]: trait.Iterator.html#method.map
 /// [`Iterator`]: trait.Iterator.html
+///
+/// # Notes about side effects
+///
+/// The [`map()`] iterator implements [`DoubleEndedIterator`], meaning that
+/// you can also [`map()`] backwards:
+///
+/// ```rust
+/// let v: Vec<i32> = vec![1, 2, 3].into_iter().rev().map(|x| x + 1).collect();
+///
+/// assert_eq!(v, [4, 3, 2]);
+/// ```
+///
+/// [`DoubleEndedIterator`]: trait.DoubleEndedIterator.html
+///
+/// However, if you call `.rev()` to iterate in the other direction, it
+/// will only be 'backwards' if your closure has no state. In other words,
+/// something like this might act in a way that you wouldn't expect:
```

Comment:
Oh, a few other things. I intentionally used a counter declared in the surrounding function, since then we don't have to use a `move` closure, and we don't have to play tricky games with state.